File: GameJam/Assets/Scripts/Inventory/Inventory.cs
using System.Collections;
using System.Collections.Generic;
using UnityEngine;
using UnityEngine.UI;

public class Inventory// : MonoBehaviour
{
    public int height;
    public int width;
    public int itemCount;

    public Cell[] cells;

    public Inventory(int ht, int wd) {
        this.height = ht;
        this.width = wd;
        this.cells = new Cell[this.height * this.width];
        this.itemCount = 0;

        for (int i = 0; i < this.height * this.width; i++) {
            this.cells[i] = new Cell();
        }
    }

    public void addItem(int idx, Item item) {
        for (int i = 0; i < item.width; i++) {
            for (int j = 0; j < item.height; j++) {
                this.cells[idx + i + (j * this.width)].item = item;
            }
        }

        this.itemCount++;
    }

    public void removeItem(Item toRemove) {
        foreach (Cell cell in this.cells)
        {
            if (cell.getItem() == toRemove) {
                cell.setItem(null);
            }
        }

        this.itemCount--;
    }

    public bool isFull() {
        foreach (Cell cell in this.cells)
        {
            if (cell.item == null) {
                return false;
            }
        }
        return true;
    }

    public bool roomAvailable(int idx, Item item) {
        if ((idx % this.width) + (item.width - 1) >= this.width || ((int)(idx / this.width)) + (item.height - 1) >= this.height) {
            return false;
        }

        for (int i = 0; i < item.width; i++) {
            for (int j = 0; j < item.height; j++) {
                if (this.cells[idx + i + (j * this.width)].item != null) {
                    return false;
                }
            }
        }

        return true;
    }

    public bool cellTaken(int idx) {
        return this.cells[idx].item != null;
    }

    public List<Item> getItems() {
        List<Item> items = new List<Item>();

        foreach (Cell cell in this.cells)
        {
            if (cell.item != null && !items.Contains(cell.item)) {
                items.Add(cell.item);
            }
        }

        return items;
    }

    public void clear() {
        foreach (Cell cell in this.cells)
        {
            cell.item = null;
        }

        this.itemCount = 0;
    }

    public void display() {
        string output = "";

        for (int i = 0; i < this.height * this.width; i++) {
            output += "| " + this.cells[i].getItemStr() + " |";
            if (i % this.width == this.width - 1) {
                output += "\n";
            }
        }

        Debug.Log(output);
    }
}

File: ProofOfConcept/inventory.py
from cell import Cell
from item import Item

import random

class Inventory:
    height: int
    width: int

    cells: list

    def __init__(self, ht, wd) -> None:
        self.height = ht
        self.width = wd
        self.cells = []

        for i in range(0, self.height * self.width):
            self.cells.append(Cell(None))

    def display(self) -> None:
        output = ""

        for i in range(0, self.width * self.height):
            output += "| " + self.cells[i].getItem() + " |"
            if i % self.width == self.width - 1:
                output += "\n"

        print(output)

    def populate(self) -> None:
        itemID = 0
        # While not full...
        while not self.isFull():
            # Get random fill point
            idx = random.randint(0, (self.height * self.width) - 1)
            if self.cells[idx].item != None:
                # If starting cell is not empty...
                continue
            else:
                # If starting cell is empty...
                # Generate random item of random size
                item = Item(itemID, random.randint(1, 3), random.randint(1, 5))

                # 50/50 chance to rotate item...
                if random.randint(0, 99) >= 50:
                    item.rotate()

                # Check if room for random item at starting point...
                if not self.roomAvailable(idx, item):
                    continue
                else:
                    itemID += 1

                    for i in range(0, item.width):
                        for j in range(0, item.height):
                            self.cells[idx + i + (j * self.width)].item = item
                    
                    self.display()

    def isFull(self) -> bool:
        for i in range(0, self.height * self.width):
            if self.cells[i].item == None:
                return False
        return True

    def roomAvailable(self, idx, item) -> bool:
        if ((idx % self.width) + (item.width - 1) >= self.width or int(idx / self.height) + (item.height - 1) >= self.height):
            return False
        
        for i in range(0, item.width):
            for j in range(0, item.height):
                if self.cells[idx + i + (j * self.width)].item != None:
                    return False

        return True